### PR TITLE
chore: upgrade to typescript 4.7 [DET-7499]

### PR DESCRIPTION
**Pull request by hamidzr**
## Description
benefit from new developments in TS including [this](https://determined-ai.slack.com/archives/CKQ3EV0G6/p1654718419336509) 
4.2 to 4.7  https://github.com/microsoft/TypeScript/compare/v4.7.3...v4.2.2

[companion shared-web pr](https://github.com/determined-ai/shared-web/pull/9)

## Commentary (optional)

```
Users/hmd/projects/da/determined/webui/react/src/hooks/useModal/useJupyterLabModal.tsx
TypeScript error in /Users/hmd/projects/da/determined/webui/react/src/hooks/useModal/useJupyterLabModal.tsx(213,75):
Object is of type 'unknown'.  TS2571

    211 |                 } catch (err) {
    212 |                   setButtonDisabled(true);
  > 213 |                   return Promise.reject(new Error(`Invalid YAML on line ${err.mark.line}.`));
        |                                                                           ^
    214 |                 }
    215 |               },
    216 |             },


make[1]: *** [build/done.stamp] Error 1
make: *** [default] Error 2


during check

WARNING: You are currently running a version of TypeScript which is not officially supported by @typescript-eslint/typescript-estree.

You may find that it works just fine, or you may not.

SUPPORTED TYPESCRIPT VERSIONS: >=3.3.1 <4.3.0

YOUR TYPESCRIPT VERSION: 4.7.3

Please only submit bug reports when using the officially supported version.

```

## Checklist

- [x] update submodule reference.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by hamidzr** [comment]
new suggestions

```
/Users/hmd/projects/da/determined/webui/react/src/pages/ExperimentDetails/ExperimentVisualization.tsx
  267:19  error  Change to an optional chain  @typescript-eslint/prefer-optional-chain
  267:20  error  Change to an optional chain  @typescript-eslint/prefer-optional-chain

/Users/hmd/projects/da/determined/webui/react/src/pages/ExperimentDetails/ExperimentVisualization/HpHeatMaps.tsx
  277:22  error  Change to an optional chain  @typescript-eslint/prefer-optional-chain

/Users/hmd/projects/da/determined/webui/react/src/pages/ExperimentDetails/ExperimentVisualization/HpScatterPlots.tsx
  181:22  error  Change to an optional chain  @typescript-eslint/prefer-optional-chain
  188:25  error  Change to an optional chain  @typescript-eslint/prefer-optional-chain

/Users/hmd/projects/da/determined/webui/react/src/shared/utils/service.ts
  10:10  error  Change to an optional chain  @typescript-eslint/prefer-optional-chain

/Users/hmd/projects/da/determined/webui/react/src/shared/utils/string.test.ts
  145:39  error  This number literal will lose precision at runtime  @typescript-eslint/no-loss-of-precision

```

File: webui/react/.eslintrc.js [hunk 1]
```diff
@@ -26,6 +26,8 @@ module.exports = {
   plugins: [ 'import', 'react', 'react-hooks', 'sort-keys-fix' ],
   root: true,
   rules: {
+    // Can disagree with @typescript-eslint/member-ordering.
+    '@typescript-eslint/adjacent-overload-signatures': 'off',
```
Comment (hamidzr) [inline webui/react/.eslintrc.js:30]: I couldn't figure out how to configure the updated member-ordering to work nicely with this rule so I turn this one off for now as I don't think it's contributing much to our codebase as is. we can turn back it back on when we configure member-ordering.
https://typescript-eslint.io/rules/member-ordering/


https://typescript-eslint.io/rules/adjacent-overload-signatures/#when-not-to-use-it very helpful :))

**Comment by hamidzr** [comment]
paused until we land https://github.com/determined-ai/determined/pull/4285

File: webui/react/src/hooks/useModal/useJupyterLabModal.tsx [hunk 2]
```diff
@@ -221,7 +221,7 @@ const JupyterLabFullConfig: React.FC<FullConfigProps> = (
                   yaml.load(value);
                   setButtonDisabled(false);
                   return Promise.resolve();
-                } catch (err) {
+                } catch (err: any) { // eslint-disable-line @typescript-eslint/no-explicit-any
```
Comment (hkang1) [inline webui/react/src/hooks/useModal/useJupyterLabModal.tsx:224]: **question:** `unknown` doesn't work huh

File: webui/react/src/pages/ExperimentDetails/ExperimentVisualization.tsx [hunk 3]
```diff
@@ -264,7 +264,7 @@ const ExperimentVisualization: React.FC<Props> = ({
 
     setFilters(prev => {
       if (prev.hParams.length !== 0) return prev;
-      const map = ((hpImportanceMap || {})[prev.metric.type] || {})[prev.metric.name];
+      const map = hpImportanceMap?.[prev.metric.type]?.[prev.metric.name] || {};
```
Comment (hkang1) [inline webui/react/src/pages/ExperimentDetails/ExperimentVisualization.tsx:267]: 👍

**Review by hkang1 — APPROVED** [review]
LGTM, thanks for the upgrades and the updates!

**Comment by hamidzr** [comment]
> I'm surprised members being alphabetical doesn't play nicely with the overload signatures, since won't that keep overloads all next to each other (which should all have the same function/member names right?)

one would want to put all getter first then setters but the other one wants to put setters and getters of the same attribute close to eachother which conflict

**Pull request merged**